reproducibility_tests/reproducibility_utils.jl

```diff
@@ -118,57 +159,156 @@ function latest_comparable_paths(;
+
+Returns a vector of reproducibility bins.
+
+Each bin is sorted such that
+`bin[1]` is the newest and `bin[end]`
+is the oldest comparable reference.
+
+```
+comparable states
+         |                           ref counter changes ---->                         | oldest
+         |                                                                             |
+         |  bin 1      bin 2      bin 3      bin 4      bin 5      bin 6      bin 7    |
+         |                                                                             |
+         |  02_49f92   04_36ebe   05_beb8a   06_4d837   05_8c311   08_45875   10_bc1e0 |
+         |             04_d6e48              06_d6d73              08_1cc58            |
+         v             04_4c042                                                        v newest
+```
+"""
+compute_bins(root_path::String) =
+    compute_bins(reverse(sorted_dataset_folder(; dir = root_path)))
+function compute_bins(sorted_paths::Vector{String})
+    isempty(sorted_paths) && return Vector{String}[]
+    bins = Vector{String}[]
+    local paths_per_bin
+    p_first_next = nothing
+    k = 1
+    while true
+        paths_per_bin = String[]
+        k > length(sorted_paths) && break
+        while true
+            k > length(sorted_paths) && break
+            p = sorted_paths[k]
+            rcp = read_ref_counter(joinpath(p, "ref_counter.jl"))
+            if isempty(paths_per_bin)
+                push!(paths_per_bin, p)
+                k += 1
             else
-                ref_counter = parse(Int, first(readlines(ref_counter_file)))
-                # Just to be safe, let's also make sure that we don't delete
-                # any paths with recent (let's say 5) ref counter increments ago.
-                if ref_counter + keep_latest_ref_counters < ref_counter_main
-                    push!(paths_to_delete, path)
+                rcb = read_ref_counter(
+                    joinpath(first(paths_per_bin), "ref_counter.jl"),
+                )
+                if rcp == rcb
+                    push!(paths_per_bin, p)
+                    k += 1
+                else
+                    break
                 end
             end
         end
-        @show ref_counter_main
-        @show length(sorted_paths)
-        @show length(paths_to_delete)
-        @info "Deleting files:"
-        for i in 1:length(paths_to_delete)
-            f = paths_to_delete[i]
-            @info "     (File, date): ($(f), $(Dates.unix2datetime(stat(f).mtime))). Reason: $(reason(f))"
-        end
-        for i in 1:length(paths_to_delete)
-            rm(paths_to_delete[i]; recursive = true, force = true)
+        push!(bins, paths_per_bin)
+    end
+    return bins
+end
+
+"""
+    get_reference_paths_to_delete(;
+        root_path,
+        keep_n_comparable_states = 5,
+        keep_n_bins_back = 7,
+    )
+
+Returns a list of folders to delete.
+
+Our reference folders are saved, and can
+therefore build up significantly.
```

Comment:
```suggestion
Our reference folders are saved on a shared folder on the cluster, and can
therefore build up significantly and take a lot of storage.
```

Reply:
So, one reason I left this generic is for re-useability / portability. I'd like to remove, or make as many things not actually depend on central (e.g., the unit tests can now exercise this locally, which doesn't have anything to do with central).

Reply:
I kept the `and take a lot of storage` piece.

Reply:
Okay, I just wanted to tell readers why we care about removing things.

Reply:
I think we can do that where it's called in `move_output.jl`